From 5d0a33783581595f800615e7dcc5c6834a9564a9 Mon Sep 17 00:00:00 2001
From: Jerome Brunet <jbrunet@baylibre.com>
Date: Mon, 13 May 2024 16:37:53 +0200
Subject: [PATCH] board: amlogic: vim3: retry mcu i2c read before failing

It is important for the DT and MCU state to be aligned.
If the MCU is in PCIe/USB2 mode and DT stays in USB3 mode, the
device is not going to operate properly.

Rarely, the i2c read fails when booting the vim3.
Retrying solved the problem so far. Limit is set to 5. Past that
we should assume that MCU has a problem and reset the device.

Signed-off-by: Jerome Brunet <jbrunet@baylibre.com>
Upstream-Status: Pending
Signed-off-by: Jerome Brunet <jbrunet@baylibre.com>
---
 board/amlogic/vim3/vim3.c | 33 ++++++++++++++++++++++++---------
 1 file changed, 24 insertions(+), 9 deletions(-)

diff --git a/board/amlogic/vim3/vim3.c b/board/amlogic/vim3/vim3.c
index 8bdfb302f728..f32b0ef7d6d3 100644
--- a/board/amlogic/vim3/vim3.c
+++ b/board/amlogic/vim3/vim3.c
@@ -15,8 +15,11 @@
 #include <asm/arch/sm.h>
 #include <asm/global_data.h>
 #include <i2c.h>
+#include <linux/delay.h>
 #include "khadas-mcu.h"
 
+#define VIM3_MCU_READ_RETRY 5
+
 int mmc_get_env_dev(void)
 {
 	switch (meson_get_boot_device()) {
@@ -41,7 +44,7 @@ int mmc_get_env_dev(void)
 int meson_ft_board_setup(void *blob, struct bd_info *bd)
 {
 	struct udevice *bus, *dev;
-	int node, i2c_node, ret;
+	int node, i2c_node, ret, i;
 	unsigned int i2c_addr;
 	u32 *val;
 
@@ -56,7 +59,7 @@ int meson_ft_board_setup(void *blob, struct bd_info *bd)
 	val = (u32 *)fdt_getprop(gd->fdt_blob, node, "reg", NULL);
 	if (!val) {
 		printf("vim3: cannot find khadas,mcu node i2c addr\n");
-		return 0;
+		return -EINVAL;
 	}
 	i2c_addr = fdt32_to_cpu(*val);
 
@@ -64,7 +67,7 @@ int meson_ft_board_setup(void *blob, struct bd_info *bd)
 	i2c_node = fdt_parent_offset(gd->fdt_blob, node);
 	if (node < 0) {
 		printf("vim3: cannot find khadas,mcu i2c node\n");
-		return 0;
+		return node;
 	}
 
 	ret = uclass_get_device_by_of_offset(UCLASS_I2C, i2c_node, &bus);
@@ -76,16 +79,28 @@ int meson_ft_board_setup(void *blob, struct bd_info *bd)
 	ret = i2c_get_chip(bus, i2c_addr, 1, &dev);
 	if (ret < 0) {
 		printf("vim3: cannot find i2c chip (%d)\n", ret);
-		return 0;
+		return ret;
 	}
 
-	/* Read USB_PCIE_SWITCH_REG */
-	ret = dm_i2c_reg_read(dev, KHADAS_MCU_USB_PCIE_SWITCH_REG);
-	if (ret < 0) {
+	/*
+	 * Read USB_PCIE_SWITCH_REG:
+	 * It is important for the MCU and DT state to be aligned but sometimes
+	 * reading the register fails, presumably because the MCU is not ready.
+	 * Retry 5 times if it happens, fail otherwise.
+	 */
+	for (i = 0; i < VIM3_MCU_READ_RETRY; i++) {
+		ret = dm_i2c_reg_read(dev, KHADAS_MCU_USB_PCIE_SWITCH_REG);
+		if (ret >= 0)
+			break;
+
 		printf("vim3: failed to read i2c reg (%d)\n", ret);
-		return 0;
+		mdelay(1000);
 	}
-	debug("MCU_USB_PCIE_SWITCH_REG: %d\n", ret);
+
+	if (ret < 0)
+		return ret;
+
+	printf("MCU_USB_PCIE_SWITCH_REG: %d\n", ret);
 
 	/*
 	 * If in PCIe mode, alter DT
-- 
2.45.2

